inductor: using dummy input to pack the linear weight for bfloat16 dynamic shape path

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106122


For the dynamic bfloat16 path, if we use plain weight, we can't call in amx path, so there use a dummy input(given a None value) to do the weight packing for better performance.

before:
```
onednn_verbose,exec,cpu,inner_product,x64:gemm:jit,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:ab:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,9.4292
```
after:
```
onednn_verbose,exec,cpu,inner_product,brgemm:avx512_core_amx_bf16,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:AB16b32a2b:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,0.35498

```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: using dummy input to pack the linear weight for bfloat16 dynamic shape path

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: using dummy input to pack the linear weight for bfloat16 dynamic shape path"



For the dynamic bfloat16 path, if we use plain weight, we can't call in amx path, so there use a dummy input(given a None value) to do the weight packing for better performance.

before:
```
onednn_verbose,exec,cpu,inner_product,x64:gemm:jit,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:ab:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,9.4292
```
after:
```
onednn_verbose,exec,cpu,inner_product,brgemm:avx512_core_amx_bf16,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:AB16b32a2b:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,0.35498

```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -848,20 +848,23 @@ def linear(match, *args, **kwargs):
+                # For bfloat dynamic shape path, using a dummy input to pack weight for a better performance.
+                packed_weight_inputs = (
+                    transpose_weight_node,
+                    None if free_symbols(batch_size) else batch_size,
```
Can we use size hint instead of `None` here?

XiaobingSuper replied:
Yes, updated.

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -848,20 +848,23 @@ def linear(match, *args, **kwargs):
-                    assert is_bf16_weight
-                    # For dynamic shape case, we need to pack weight in runtime.
-                    packed_weight_node = transpose_weight_node
+                if not is_bf16_weight:
+                    assert not free_symbols(batch_size)
```
Is it easier to read?
```suggestion
                if free_symbols(batch_size):
                    assert is_bf16_weight, f"only bf16 weight prepacking supports dynamic shape inputs but got {weight.meta.get("val").dtype}"
```

XiaobingSuper replied:
Changed.

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -846,22 +846,30 @@ def linear(match, *args, **kwargs):
+                    assert (
+                        is_bf16_weight
+                    ), f"only bf16 weight prepacking supports dynamic shape inputs but got {weight_dtype}"
+                # For bfloat dynamic shape path, using a dummy input to pack weight for a better performance.
```
```suggestion
                # For bfloat16 dynamic shape path, using input size hint to pack weight for a better performance.
```

XiaobingSuper committed:
Update on "inductor: using dummy input to pack the linear weight for bfloat16 dynamic shape path"



For the dynamic bfloat16 path, if we use plain weight, we can't call in amx path, so there use a dummy input(given a None value) to do the weight packing for better performance.

before:
```
onednn_verbose,exec,cpu,inner_product,x64:gemm:jit,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:ab:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,9.4292
```
after:
```
onednn_verbose,exec,cpu,inner_product,brgemm:avx512_core_amx_bf16,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:AB16b32a2b:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,0.35498

```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper committed:
Update on "inductor: using dummy input to pack the linear weight for bfloat16 dynamic shape path"



For the dynamic bfloat16 path, if we use plain weight, we can't call in amx path, so there use a dummy input(given a None value) to do the weight packing for better performance.

before:
```
onednn_verbose,exec,cpu,inner_product,x64:gemm:jit,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:ab:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,9.4292
```
after:
```
onednn_verbose,exec,cpu,inner_product,brgemm:avx512_core_amx_bf16,forward_training,src_bf16::blocked:ab:f0 wei_bf16::blocked:AB16b32a2b:f0 bia_bf16::blocked:a:f0 dst_bf16::blocked:ab:f0,attr-scratchpad:user ,,mb64ic256oc256,0.35498

```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
